Geodetic levelling

Aslak-Meister opened the pull request:
Seneste version til review primo primo januar

Commit:
Initial commit: fire/api/geodetic_levelling added.

Commit:
Grid-filer til brug for geodætisk korrektion af nivellement tilføjet.

Commit:
New environment file environment-dev-geod-lev.yml with packages necessary for geodetic_levelling added.

Commit:
Imports in modules in fire/api/geodetic_levelling updated.

Commit:
Function apply_geodetic_corrections_to_height_diffs modified to take a list of InternNivObservation as input.

Commit:
Parametre-attribut tilføjet regnemotorer.

Commit:
Regnemotoren GeodætiskRegn(GamaRegn) defineret.

Alle parametre rapporteret af metoden parametre udskrives nu i Parametre-fanebladet (i output-regnearket).

Commit:
Function apply_geodetic_corrections_to_height_diffs() slightly modified (parameter names changed).

Commit:
Function convert_geopotential_heights_to_metric_heights() modified to take a list of InternKote objects as input.

Commit:
environment-dev-geod-lev.yml opdateret som følge af rebase

Commit:
Functions apply_geodetic_corrections_to_height_diffs() and convert_geopotential_heights_to_metric_heights()
slightly modified due to rebase (new name for dataclasses for levelling observation and levelled height).

Commit:
Funktionalitet til geodætisk korrektion af nivellement samt konvertering af højder tilføjet til
regnemotor GeodætiskRegn.

Commit:
Funktionalitet til skrivning af excel-fil med anvendte geodætiske korrektioner m.m. tilføjet til GeodætiskRegn.

Commit:
Kontroller af parametre vedr. geodætiske korrektioner tilføjet til regnemotoren GeodætiskRegn samt til funktionen
apply_geodetic_corrections_to_height_diffs m.v.

Commit:
Function decimalyear_to_datetime() added to module histogram.py.

Commit:
Diverse mangler vedr. tolkning af regneparametre fra kommandolinjen og initialisering af parametre i regnemotoren GeodætiskRegn udbedret.

Commit:
Ny regnemotor DVR90Regn tilføjet.

Commit:
Dokumentation/docstring vedr. regnemotorerne GeodætiskRegn og DVR90Regn tilføjet.

krebslw commented on src/fire/api/geodetic_levelling/tidal_transformation.py:
```diff
@@ -0,0 +1,541 @@
+        exit(
+            "Function apply_tidal_corrections_to_height_diff: Wrong arguments for\n\
+        parameter tidal_system and/or gravitymodel and/or grid_inputfolder."
+        )
```
Har tit fået at vide, at det er smart at flytte den her type "exit"-betingelser op i toppen af funktionen, så man checker særtilfældene "up front" om funktionen skal fejle. Fx kunne der helt i starten stå:
```
    allowed_tidal_systems = ["non", "mean", "zero"]
    if not tidal_system in allowed_tidal_systems:
        raise ValueError(f"`tidal_system` must be one of {', '.join(allowed_tidal_systems)}")

    if tidal_system != "non" and ((gravitymodel is None) or (grid_inputfolder is None)):
        exit(
            "Function apply_tidal_corrections_to_height_diff: Wrong arguments for\n\
        parameter gravitymodel and/or grid_inputfolder."
        )
```

Aslak-Meister replied:
OK, det er implementeret.

krebslw commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+
+        point_to_gravity = transform_gravity_from_tidal_system_to_tidal_system(
+            point_to_gravity, point_to_lat, "zero_to_mean"
+        )
```
Her kan der spares et par linjer med:
```suggestion
    if tidal_system != "zero":
        if tidal_system == "non":
            transformation = "zero_to_non"
        elif tidal_system == "mean":
            transformation = "zero_to_mean"

        point_from_gravity = transform_gravity_from_tidal_system_to_tidal_system(
            point_from_gravity, point_from_lat, transformation
        )
        point_to_gravity = transform_gravity_from_tidal_system_to_tidal_system(
            point_to_gravity, point_to_lat, transformation
        )
```

Aslak-Meister replied:
OK, det er implementeret.

krebslw commented on src/fire/api/geodetic_levelling/contour_plot.py:
```diff
@@ -0,0 +1,171 @@
+"""This module contains functions for the generation of contour plots showing
```
Modul til kontur-plot bruges ikke nogen steder. Mener ikke det bør slettes, men vi kan overveje at tage det ud her i første omgang af pull requestet, indtil det finder sin rigtige plads hvor det skal bruges?

Aslak-Meister replied:
OK, jeg har gjort modulet untracked.

krebslw commented on src/fire/api/geodetic_levelling/geodetic_correction_levelling_obs.py:
```diff
@@ -0,0 +1,215 @@
+            exit(
+                "Function apply_geodetic_corrections_to_height_diffs: Wrong arguments for\n\
+            parameter epoch_target and/or deformationmodel and/or grid_inputfolder."
+            )
```
Er det nødvendigt at smide den her fejl? Som jeg forstår det, så skal der kun laves uplift-korrektion hvis både `epoch_target`,  `deformationmodel` og `grid_inputfolder` er givet. Hvis bare én af disse er None, så skal der ikke laves uplift korrektion, og vi kan fortsætte til næste trin.

krebslw commented on src/fire/api/geodetic_levelling/geodetic_correction_levelling_obs.py:
```diff
@@ -0,0 +1,215 @@
+                height_diff_object.id,
+                f"ΔH tidal correction (tidal system: {tidal_system}) [m]",
+            ] = tidal_corr
+
```
Da funktionen ikke skal gøre mere herefter hvis `grid_inputfolder` er None kunne man her indsætte :
```
    if grid_inputfolder is None:
        return height_diff, corrections
```
hvilket gør at man nedenunder kan fjerne tjek for om `grid_inputfolder` er None

Aslak-Meister replied:
Jeg synes ikke umiddelbart, at dette er en forbedring. Koden bliver en anelse længere og en anelse mere kompliceret. Og alt efter hvilke korrektioner, der skal udføres, enten sparer man én eller får én ekstra betingelsessætning, der skal evalueres.

krebslw reviewed (changes requested):
Overordnet synes jeg det ser virkeligt godt ud, og det virker som det skal "ud-af-boksen". 
Din kodestil er meget stringent og altid godt med gode docstrings :) Og synes du er lykkedes med at integrere det mere eller mindre sømløst med regnemotor-modulet.

De fleste af mine kommentarer retter sig til logikken omkring "kontrol-flowet" hvor jeg mener man kan spare en del kompleksitet. Men jeg er selv god til at finde på kringlede logiske konstruktioner i min kode, så tag ikke det hele som den absolutte sandhed :)

Dét jeg helst så rettet inden vi kan merge det, er at få dataframes helt ud af geodetic-levelling-pakken. De underliggende funktioner bør returnere de nødvendige størrelser, som det overliggende lag (regnemotor) kan pakke sammen til en dataframe og som så gemmes som excel.

Commit:
Optional use of approx formulas (instead of rigorous formulas) for tidal transformation of gravity and height differences implemented.

Commit:
Function calculate_perm_tidal_deformation_geoid() modified: now gravity used for calculation of
the permanent tidal deformation of the geoid is transformed to non-tidal system.

Commit:
Koden vedr. fortolkning af regneparametre er flyttet til en særskilt funktion fortolk_regneparametre()

Aslak-Meister commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+        # The iteration is started if height_converted is not nan
+        if iterate == True and isnan(height_converted) == False:
+            while not (
+                -0.00001 <= (height_converted - approx_normal_height) <= 0.00001
```
Det tror jeg ikke, at en bruger har behov for. 

Til gengæld kan man spare en iteration ved at slække på stopkriteriet, uden at det har nogen praktisk betydning for de konverterede højder. Jeg har blot ikke endnu beregnet, hvad stopkriteriet bør være, og denne forbedring kommer først med i næste version.

Commit:
Metoden læs_gama_outputfil(self) let modificeret: Nord og øst kopieres nu fra self._gamle_koter til NivKote-objekter i output-liste.

Commit:
Functions for interpolation/calculation of gravity etc. moved to new module gravity.py.

Aslak-Meister commented on src/fire/api/geodetic_levelling/tidal_transformation.py:
```diff
@@ -0,0 +1,541 @@
+
+from fire.api.geodetic_levelling.metric_to_gpu_transformation import (
+    interpolate_gravity,
+)
```
Jeg havde placeret importen nederst i tidal_transformation.py for at undgå cirkulær import. Nu har jeg i stedet placeret tyngderelaterede funktioner, herunder interpolate_gravity, i et nyt selvstændigt modul gravity.py.

Aslak-Meister commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+        exit(
+            "Function convert_geopotential_height_to_normal_height: Wrong argument for\n\
+        parameter conversion."
+        )
```
OK, det er implementeret.

Aslak-Meister commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+        exit(
+            "Function convert_geopotential_height_to_helmert_height: Wrong argument for\n\
+        parameter conversion."
+        )
```
OK, det er implementeret.

Aslak-Meister commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+    ?
+    """
+    # Conversion of a geopotential height to normal height
+    if conversion == "geopot_to_normal":
+        # Calculation of average normal gravity in units of 10 m/s^2
+        average_normal_gravity = (
+            calculate_average_normal_gravity(latitude, approx_normal_height) * 0.1
+        )
+
+        # Normal height in units of meters
+        height_converted = height / average_normal_gravity
+
+        # Iterative calculation of normal height
+        # The iteration is started if height_converted is not nan
+        if iterate == True and isnan(height_converted) == False:
+            while not (
+                -0.00001 <= (height_converted - approx_normal_height) <= 0.00001
+            ):
+                approx_normal_height = height_converted
+
+                # Calculation of average normal gravity in units of 10 m/s^2
+                average_normal_gravity = (
+                    calculate_average_normal_gravity(latitude, approx_normal_height)
+                    * 0.1
+                )
+
+                # Normal height in units of meters
+                height_converted = height / average_normal_gravity
+
+    # Conversion of a normal height to geopotential height
+    elif conversion == "normal_to_geopot":
+        # Calculation of average normal gravity in units of 10 m/s^2
+        average_normal_gravity = (
+            calculate_average_normal_gravity(latitude, height) * 0.1
+        )
+
+        # Geopotential height in units of gpu (1 gpu = 10 m^2/s^2)
+        height_converted = height * average_normal_gravity
```
Jeg mener, at dette ændringsforslag gør koden lidt mindre læsbar, fordi det bliver mere implicit, hvad koden gør.

Aslak-Meister commented on src/fire/api/geodetic_levelling/metric_to_gpu_transformation.py:
```diff
@@ -0,0 +1,603 @@
+    # If Helmert heights or normal heights are to be converted to geopotential heights
+    # the input heights are taken from column "Kote" in the sheet "Kontrolberegning" in the
+    # input excel-file and the converted heights are written to column "Ny kote" in the
+    # sheet "Kontrolberegning" in the output excel-file.
```
Docstringen er blevet omskrevet.

Commit:
Minor improvements to module metric_to_gpu_transformation.py regarding check of input arguments etc.

Aslak-Meister commented on src/fire/cli/niv/_regn.py:
OK, det er implementeret.

Aslak-Meister commented on src/fire/api/niv/regnemotor.py:
```diff
@@ -527,6 +564,349 @@ def filer(self) -> list:
+                    (gammel_kote.nord, gammel_kote.øst)
+                    for gammel_kote in self.gamle_koter
+                    if gammel_kote.punkt == punktnr
+                ][0]
```
OK, jeg har valgt at ændre læs_gama_outputfil.

Aslak-Meister commented on src/fire/api/niv/regnemotor.py:
OK, det er implementeret.

Aslak-Meister commented on src/fire/api/geodetic_levelling/tidal_transformation.py:
```diff
@@ -0,0 +1,541 @@
+            gravitymodel,
+        )
+
+        tidal_corr = height_diff_corrected - height_diff
```
OK, det er implementeret.

Commit:
Module contour_plot.py removed from tracking.

Commit:
Module histogram.py removed from tracking and as a consequence hereof time conversion in regnemotor.py slightly modified.

Aslak-Meister commented on src/fire/api/niv/regnemotor.py:
```diff
@@ -527,6 +564,349 @@ def filer(self) -> list:
+            # ang. "dubious year", hvilke vi ikke ønsker at se i denne specifikke sammenhæng
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=erfa.ErfaWarning)
+                self.epoch_target = decimalyear_to_datetime(epoch_target)
```
Ok, implementeret.

Aslak-Meister commented on src/fire/api/geodetic_levelling/histogram.py:
```diff
@@ -0,0 +1,119 @@
+"""This module contains functions for generation of histograms visualizing
+the temporal distribution of levelling observations in a FIRE project.
+"""
+
```
OK, jeg har gjort modulet untracked.

Aslak-Meister commented on src/fire/api/geodetic_levelling/geodetic_correction_levelling_obs.py:
```diff
@@ -0,0 +1,215 @@
+                f"ΔH epoch correction (target epoch: {epoch_target}) [m]",
+            ] = epoch_corr
+
+        elif epoch_target is not None:
+            exit(
+                "Function apply_geodetic_corrections_to_height_diffs: Wrong arguments for\n\
+            parameter epoch_target and/or deformationmodel and/or grid_inputfolder."
+            )
```
Jeg har lavet det således aht. brugeren. Hvis brugeren har angivet en epoke, må det antages, at brugeren ønsker at tidspropagere observationerne, og at det derfor er en fejl, at deformationmodel og/eller grid_inputfolder ikke er angivet.

Commit:
Koden vedr. initialisering af parametre til regnemotorerne GeodætiskRegn og DVR90Regn forsimplet.

Aslak-Meister commented on src/fire/api/niv/regnemotor.py:
```diff
@@ -527,6 +564,349 @@ def filer(self) -> list:
+        if not tidal_system:
+            self.tidal_system = None
+        else:
+            self.tidal_system = tidal_system
```
OK, det er implementeret.

Aslak-Meister commented on src/fire/api/niv/regnemotor.py:
```diff
@@ -527,6 +564,349 @@ def filer(self) -> list:
+        if not grid_inputfolder:
+            self.grid_inputfolder = None
+        else:
+            self.grid_inputfolder = Path(grid_inputfolder)
```
OK, det er implementeret.

Commit:
Som følge af rebase er environment-dev.yml opdateret med nye pakker nødvendige aht. nyregnemotor.

Commit:
Bug fix: Import af pandas.io.formats.style tilføjet i src/fire/cli/niv/__init__.py (pandas.io.formats.style importeres ikke
med import pandas as pd).

Commit:
environment-dev-geod-lev.yml removed from tracking.

Commit:
environment.yml opdateret med nye pakker nødvendige aht. nyregnemotor. Derudover er environment-dev.yml nu sorteret korrekt alfabetisk.

Aslak-Meister commented on environment-dev-geod-lev.yml:
```diff
@@ -0,0 +1,38 @@
+name: fire-dev-geod-lev
```
OK, det er implementeret.

Commit:
Module geodetic_correction_levelling_obs.py modified: function for geodetic correction of height differences split into two:
one function taking a single height difference as an input (and without any use of DataFrames) and another function taking a
dict of NivObservation objects as input. Validation of input arguments moved to the beginnig of function definitions.

Aslak-Meister commented on src/fire/api/geodetic_levelling/geodetic_correction_levelling_obs.py:
```diff
@@ -0,0 +1,215 @@
+            exit(
+                "Function apply_geodetic_corrections_to_height_diffs: Wrong arguments for\n\
+            parameter height_diff_unit and/or gravitymodel and/or grid_inputfolder."
+            )
```
Forslaget er ikke længere aktuelt efter jeg har flyttet tjek af argumenter op i toppen af funktionen samt ændret logikken i tjekkene.

Aslak-Meister commented on src/fire/api/geodetic_levelling/geodetic_correction_levelling_obs.py:
```diff
@@ -0,0 +1,215 @@
+    for height_diff_object in height_diff_objects:
+        index.append(height_diff_object.id)
+
+    corrections_df = pd.DataFrame(
```
OK, jeg har splittet funktionen til geodætisk korrektion af nivellementsobservationer i to, én funktion der tager én observation som input (og ikke gør brug af dataframes) og én funktion der tager en dict af observationer.

Commit:
Testing of functions in api/geodetic_levelling implemented.
